### PR TITLE
Multiple types per question

### PR DESCRIPTION
**Pull request by Ferdinand737**
# Description

Includes all changes from [Queue Refactor PR](https://github.com/ubco-db/office-hours/pull/54)

- Professors/TAs can now create many question types
- Students can select any of the created question types

**Question types temporarily disabled for async question queue.**
- Async questions will be reworked as well but this PR is already big

New PR because Queue refactor was merged

### CODE REVIEW

**Comment by AdamFipke** [comment]
might need to solve the merge conflicts first

File: packages/app/components/Queue/Student/QuestionForm.tsx [hunk 1]
```diff
@@ -219,28 +231,40 @@ export default function QuestionForm({
...
+            >
+              {questionsTypeState.map((type) => (
+                <Select.Option value={type.id} key={type.id}>
+                  {type.name}
+                </Select.Option>
+              ))}
+            </Select>
+          </>
+        ) : (
+          <></>
```
Comment (d3li0n) [inline packages/app/components/Queue/Student/QuestionForm.tsx:266]: Should we keep for this case?
```
<p>Loading..</p>
```

File: packages/app/components/Queue/Student/QuestionForm.tsx [hunk 2]
```diff
@@ -283,3 +307,8 @@ export default function QuestionForm({
     </Modal>
   )
 }
+
+QuestionForm.propTypes = {
```
Comment (d3li0n) [inline packages/app/components/Queue/Student/QuestionForm.tsx:311]: Can we move it to the top of the file?

File: packages/app/components/Queue/TA/TAAddStudent.tsx [hunk 3]
```diff
@@ -164,35 +177,47 @@ export function AddStudentsModal({
...
+          >
+            {questionsTypeState.map((type) => (
+              <AntdSelect.Option value={type.id} key={type.id}>
+                {type.name}
+              </AntdSelect.Option>
+            ))}
+          </AntdSelect>
+        </>
+      ) : (
+        <></>
```
Comment (d3li0n) [inline packages/app/components/Queue/TA/TAAddStudent.tsx:219]: Should we keep this?

```
<p>There are No Question Types</p>
```

File: packages/app/components/Queue/TA/TAAddStudent.tsx [hunk 4]
```diff
@@ -268,3 +293,6 @@ export function AddStudentsModal({
     </Modal>
   )
 }
+AddStudentsModal.propTypes = {
```
Comment (d3li0n) [inline packages/app/components/Queue/TA/TAAddStudent.tsx:296]: Can we move it to the top of the file?

File: packages/server/migration/1702234882706-rename-question-type-col.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class renameQuestionTypeCol1702234882706 implements MigrationInterface {
+  name = 'renameQuestionTypeCol1702234882706';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
```
Comment (d3li0n) [inline packages/server/migration/1702234882706-rename-question-type-col.ts:7]: My concern is why this migration didn't move to new migration files? Do we need any of these fields?

```
 await queryRunner.query(
      `ALTER TABLE "course_model" DROP CONSTRAINT "FK_e04e8cfa5bc6ca85981181a9119"`,
    );
    await queryRunner.query(
      `ALTER TABLE "course_model" DROP COLUMN "organizationCourseId"`,
    );
    await queryRunner.query(
      `ALTER TABLE "organization_course_model" ADD CONSTRAINT "UQ_4fef22be04e7b58e8728a24b207" UNIQUE ("courseId")`,
    );
    await queryRunner.query(
      `ALTER TABLE "organization_course_model" ADD CONSTRAINT "FK_4fef22be04e7b58e8728a24b207" FOREIGN KEY ("courseId") REFERENCES "course_model"("id") ON DELETE NO ACTION ON UPDATE NO ACTION`,
    );
```

File: packages/server/src/question/question.controller.ts [hunk 6]
```diff
@@ -493,30 +508,19 @@ export class QuestionController {
...
       return;
     }
-    const strQ: string[] = [];
-    for (let i = 0; i < questions.length; i++) {
-      strQ[i] = questions[i].question;
-    }
-    res.status(200).send(strQ);
-    return strQ;
+    res.status(200).send(questions);
+    return questions;
```
Comment (d3li0n) [inline packages/server/src/question/question.controller.ts:512]: Do we need a return statement here? I think res should return it?

File: packages/server/src/seed/seed.controller.ts [hunk 7]
```diff
@@ -389,12 +390,26 @@ export class SeedController {
...
+        ...options,
+        queueId: body.queueId,
+        text: body.data.text,
+        creatorId: body.studentId,
+        groupable: body.data.groupable,
+        location: body.data.location,
+        questionTypes: questionTypes,
+        createdAt: new Date(),
+      });
+      return question;
```
Comment (d3li0n) [inline packages/server/src/seed/seed.controller.ts:409]: I don't think we need to return anything? Also should we use `.save()` as well here?

File: packages/server/test/question.integration.ts [hunk 8]
```diff
@@ -128,16 +134,26 @@ describe('Question Integration', () => {
         allowQuestions: true,
         staffList: [ta.user],
       });
+      const questionTypes = [];
+      forEach(QuestionTypes, async (questionType) => {
+        const currentQuestionType = await QuestionTypeFactory.create({
+          name: questionType.name,
+          color: questionType.color,
+          cid: course.id,
+        });
```
Comment (d3li0n) [inline packages/server/test/question.integration.ts:143]: Shouldn't we save them as well in database?

**Review by d3li0n — CHANGES_REQUESTED** [review]
Left some comments

File: packages/server/src/seed/seed.controller.ts (hunk 7 above)
Comment (Ferdinand737) [reply packages/server/src/seed/seed.controller.ts:409]: seeds controller had a bunch of unused code, I have deleted it.

File: packages/server/migration/1702234882706-rename-question-type-col.ts (hunk 5 above)
Comment (Ferdinand737) [reply packages/server/migration/1702234882706-rename-question-type-col.ts:7]: I am not sure what you mean.

**Review by wskksw — COMMENTED** [review]
Overall, I have some usability recommendations regarding the component of adding question types as a professor. 
Currently, there are lot's options/ways to pick colors, but is that necessary? Maybe just provide up to 10 default colors for questions, or make it a collapsable component to pick complex colors? I am not sure about this one just recommendation. 
The other thing is that there is no margin between types and input or the color pallette and add. 
image

**Review by AdamFipke — CHANGES_REQUESTED** [review]
I'm pretty happy with it though I noticed there's still the bug of being able to create types with no type name.

Steps: 
1. Create a question type
2. Click 'add' with the question type text box empty. This seems to bypass the error checking and create an empty question type that can't be removed

**Comment by AdamFipke** [comment]
wait I also found another critical bug I think: the question types are the same across all queues. 

Steps:
1. Create a question type.
2. Move to a different queue and edit queue details. You should see the question type you just made even though you're in a new queue

**Comment by Ferdinand737** [comment]
> I'm pretty happy with it though I noticed there's still the bug of being able to create types with no type name.
> 
> Steps:
> 
> 1. Create a question type
> 2. Click 'add' with the question type text box empty. This seems to bypass the error checking and create an empty question type that can't be removed

Screenshot from 2024-01-19 15-14-23

I Cannot replicate this issue. There is error handling for this case.

**Comment by Ferdinand737** [comment]
> wait I also found another critical bug I think: the question types are the same across all queues.
> 
> Steps:
> 
> 1. Create a question type.
> 2. Move to a different queue and edit queue details. You should see the question type you just made even though you're in a new queue

This was the case with the existing system. This is how it is on master. I could change it. But Profs might want question types shared.

**Comment by AdamFipke** [comment]
> > wait I also found another critical bug I think: the question types are the same across all queues.
> > Steps:
> > 
> > 1. Create a question type.
> > 2. Move to a different queue and edit queue details. You should see the question type you just made even though you're in a new queue
> 
> This was the case with the existing system. This is how it is on master. I could change it. But Profs might want question types shared.

Oh I see. Okay maybe we can discuss this at a later date then, should be good for this PR

**Review by AdamFipke — APPROVED** [review]
LGTM

**Review by wskksw — APPROVED** [review]

**Pull request merged**